server/src/lib/stats.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 Oxide Computer Company
+use anyhow::anyhow;
+use dropshot::{ConfigDropshot, ConfigLogging, ConfigLoggingLevel};
+use omicron_common::api::internal::nexus::ProducerEndpoint;
+use oximeter::{
+    types::{Cumulative, Sample},
+    Metric, MetricsError, Producer, Target,
+};
+use oximeter_producer::{Config, Server};
+use slog::{error, info, Logger};
+
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+use uuid::Uuid;
+
+// These structs are used to construct the desired stats for Oximeter.
+#[derive(Debug, Copy, Clone, Target)]
+struct InstanceUuid {
+    pub uuid: Uuid,
+}
+#[derive(Debug, Default, Copy, Clone, Metric)]
+pub struct Rebooted {
+    /// Count of times instance was rebooted
+    #[datum]
+    pub count: Cumulative<i64>,
+}
+
+// All the counter stats in one struct.
+// To create additional stats that Oximeter will collect, add fields to this
+// structure.  See Oximeter for details, but each fields should be
+// constructed similar to "run_count".  In addition, a new method should
+// be added to the PropStatOuter impl that will be called when the new
+// field has changed.
+#[derive(Clone, Debug)]
+pub struct PropCountStat {
+    stat_name: InstanceUuid,
+    run_count: Rebooted,
+}
+
+impl PropCountStat {
+    pub fn new(uuid: Uuid) -> Self {
+        PropCountStat {
+            stat_name: InstanceUuid { uuid: uuid },
+            run_count: Default::default(),
+        }
+    }
+    pub fn uuid(&self) -> Uuid {
+        return self.stat_name.uuid;
+    }
+}
+
+// This struct wraps the stat struct in an Arc/Mutex so the worker tasks can
+// share it with the producer trait.
+#[derive(Clone, Debug)]
+pub struct PropStatOuter {
+    pub prop_stat_wrap: Arc<Mutex<PropCountStat>>,
+}
+
+impl PropStatOuter {
+    // When an operation happens that we wish to record in Oximeter,
+    // one of these methods will be called.  Each method will get the
+    // correct field of PropCountStat to record the update.
+    pub fn add_activation(&self) {
+        let mut pso = self.prop_stat_wrap.lock().unwrap();
+        let datum = pso.run_count.datum_mut();
+        *datum += 1;
+    }
+}
+
+// This trait is what is called to update the data to send to Oximeter.
+// It is called on whatever interval was specified when setting up the
+// connection to Oximeter.  Since we get a lock in here (and on every
+// IO, don't call this too frequently, for some value of frequently that
+// I'm not sure of.
```

Comment:
Nit: might be good to have a `const OXIMETER_STATS_INTERVAL` somewhere with the warning about querying too frequently nearby (or just move this down by line 126 where the interval is provided).

Reply:
Updated in 395d4db.
I tried to improve the comments a bit more as well.